Mathlib/RingTheory/Ideal/Over.lean

```diff
@@ -430,3 +430,98 @@ lemma map_eq_top_iff {R S} [CommRing R] [CommRing S]
+a maximal ideal. -/
+instance Ideal.IsMaximal.under (A : Type*) [CommRing A] {B : Type*} [CommRing B] (P : Ideal B)
+    [P.IsMaximal] [Algebra A B] [Algebra.IsIntegral A B] : (Ideal.under A P).IsMaximal :=
+  isMaximal_comap_of_isIntegral_of_isMaximal P
+
+variable {A}
+
+/-- `P` lies over `p` if `p` is the preimage of `P` of the `algebraMap`. -/
+class ideal_lies_over : Prop where
+  over : p = Ideal.comap (algebraMap A B) P
+
+@[inherit_doc] infix : 50 "lies_over" => ideal_lies_over
+
+instance Ideal.over_under : P lies_over (Ideal.under A P) where over := rfl
+
+variable {A : Type*} [CommSemiring A] (p : Ideal A) {B : Type*} [CommSemiring B] (P : Ideal B)
+  {C : Type*} [Semiring C] (𝔓 : Ideal C) [Algebra A B] [Algebra B C] [Algebra A C]
+  [IsScalarTower A B C]
+
+theorem ideal_lies_over.trans [hp : 𝔓 lies_over P] [hP : P lies_over p] : 𝔓 lies_over p where
+  over := by rw [hP.over, hp.over, Ideal.comap_comap, ← IsScalarTower.algebraMap_eq]
+
+namespace Ideal
+
+theorem lies_over_tower_bot [hp : 𝔓 lies_over p] [hP : 𝔓 lies_over P] : P lies_over p where
+  over := by rw [hp.over, hP.over, comap_comap, ← IsScalarTower.algebraMap_eq]
+
+variable (B)
+
+instance under_lies_over_of_lies_over [𝔓 lies_over p] : (Ideal.under B 𝔓) lies_over p :=
+  lies_over_tower_bot p (Ideal.under B 𝔓) 𝔓
+
+theorem eq_under_under_of_lies_over [𝔓 lies_over p] : p = Ideal.under A (Ideal.under B 𝔓) :=
+  (under_lies_over_of_lies_over p B 𝔓).over
+
+variable {A : Type*} [CommRing A] (p : Ideal A) {B : Type*} [CommRing B] (P : Ideal B)
+  [Algebra A B] [Algebra.IsIntegral A B] [ho : P lies_over p]
+
+theorem isMaximal_of_lies_over_isMaximal [hpm : p.IsMaximal] [P.IsPrime] : P.IsMaximal := by
+  rw [ho.over] at hpm
+  exact isMaximal_of_isIntegral_of_isMaximal_comap P hpm
+
+theorem isMaximal_of_isMaximal_lies_over [P.IsMaximal] : p.IsMaximal := by
+  rw [ho.over]
+  exact isMaximal_comap_of_isIntegral_of_isMaximal P
+
+/-- If `P` lies over `p`, then canonically `B ⧸ P` is a `A ⧸ p`-algebra. -/
+instance algebra_quotient_of_lies_over : Algebra (A ⧸ p) (B ⧸ P) :=
+  Ideal.Quotient.algebraQuotientOfLEComap (le_of_eq ho.over)
+
+instance : IsScalarTower A (A ⧸ p) (B ⧸ P) :=
+  IsScalarTower.of_algebraMap_eq (fun _ ↦ rfl)
+
+/-- `B ⧸ P` is a finite `A ⧸ p`-module if `B` is a finite `A`-module. -/
+instance module_finite_of_lies_over [Module.Finite A B] : Module.Finite (A ⧸ p) (B ⧸ P) :=
+  Module.Finite.of_restrictScalars_finite A (A ⧸ p) (B ⧸ P)
+
+example [Module.Finite A B] : Module.Finite (A ⧸ comap (algebraMap A B) P) (B ⧸ P) := inferInstance
+
+variable [p.IsMaximal] (B : Type*) [CommRing B] [Nontrivial B] [Algebra A B]
+  [NoZeroSMulDivisors A B] [Algebra.IsIntegral A B]
+
+/-- For any maximal idela `p` in `A`, there exists a maximal ideal in `B` lying over `p`. -/
+noncomputable def over_isMaximal : Ideal B :=
+  Classical.choose <| exists_ideal_over_maximal_of_isIntegral p <|
+    (NoZeroSMulDivisors.algebraMap_ker_eq_bot A B).trans_le bot_le
```

Comment:
I wonder what you need this for?

Reply:
I came up with this definition while writing about Galois extensions of number fields. 
Let `K / L` be a Galois extension of number fields, and let `p` be a maximal ideal of `𝓞 K`. Initially, I defined the ramification index of `p` as `ramificationIdx (algebraMap (𝓞 K) (𝓞 L)) p P`, where `P` is a maximal ideal over `p` selected using `exists_ideal_over_maximal_of_isIntegral`. However, when I was proving that it is equal to the ramification index corresponding to any other ideal `P'`, I had to repeatedly prove the existence of `P` and then use `choose_spec` to obtain the corresponding property. 
Therefore, I wanted to use this definition to simplify my previous proof.

Reply:
You should prove that `ramificationIdx (algebraMap (𝓞 K) (𝓞 L)) p P = ramificationIdx (algebraMap (𝓞 K) (𝓞 L)) p P'` for all `P` and `P'` by some transitivity argument and the lemma you mentioned is then just a simple application of this.
Also see how this was done in [flt-regular](https://github.com/leanprover-community/flt-regular/blob/82763ff91a32d6ab0e194eb44e05b02fb8aa361b/FltRegular/NumberTheory/GaloisPrime.lean#L225-L282). 

I feel like this definition is very non-canonical and could mislead other people into providing weaker statements later on and I'd suggest we avoid it unless it gives a significant improvement.